modules/effects/src/concat_latest_from.ts

```diff
@@ -0,0 +1,38 @@
+import { Observable, of, OperatorFunction } from 'rxjs';
+import { concatMap, withLatestFrom } from 'rxjs/operators';
+
+type TypeOfObservable<T> = T extends Observable<infer U> ? U : never;
+
+/**
+ * 'concatLatestFrom' combines the source value
+ * and the last available value from a lazily evaluated Observable
+ * in a new array
+ */
+export function concatLatestFrom<T extends Observable<unknown>[], V>(
+  observablesFactory: (value: V) => [...T]
+): OperatorFunction<V, [V, ...{ [i in keyof T]: TypeOfObservable<T[i]> }]>;
+export function concatLatestFrom<T extends Observable<unknown>, V>(
+  observableFactory: (value: V) => T
+): OperatorFunction<V, [V, TypeOfObservable<T>]>;
+export function concatLatestFrom<
+  T extends Observable<unknown>[] | Observable<unknown>,
+  V
+>(
+  observablesFactory: (value: V) => T extends Observable<unknown>[] ? [...T] : T
+) {
+  return concatMap((value) => {
+    const observables = observablesFactory(value);
+    const observablesAsArray = Array.isArray(observables)
+      ? observables
+      : [observables];
+    return of(value).pipe(withLatestFrom(...observablesAsArray));
+  }) as OperatorFunction<
+    V,
+    [
+      V,
+      ...(T extends Observable<unknown>[]
+        ? { [i in keyof T]: TypeOfObservable<T[i]> }
+        : [TypeOfObservable<T>])
+    ]
+  >;
+}
```

Comment:
That casting at the end looks very overwhelming 😬
How about specifying the return type of the function, and then do a gentler cast? 🙂
 
```suggestion
export function concatLatestFrom<
  T extends Observable<unknown>[] | Observable<unknown>,
  V,
  R = [
    V,
    ...(T extends Observable<unknown>[]
      ? { [i in keyof T]: ObservedValueOf<T[i]> }
      : [ObservedValueOf<T>])
  ]
>(observablesFactory: (value: V) => T): OperatorFunction<V, R> {
  return concatMap((value) => {
    const observables = observablesFactory(value);
    const observablesAsArray = Array.isArray(observables)
      ? observables
      : [observables];
    return of(value).pipe(
      withLatestFrom(...observablesAsArray)
    ) as Observable<R>;
  });
}
```

Reply:
Yeah I was definitely at the beginning of understanding these more complex typings when I wrote this. Thanks!

Reply:
Yeah, there are some interesting types here 🙂, e.g. `{ [i in keyof T]: ObservedValueOf<T[i]> }`.
What sucks is that TS doesn't let you assign it to a type and spread it elsewhere, even though it knows that it's still an `Array`-like type - very frustrating 😐